pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1152 @@
+			.map_err(|e| {
+				// Inner error holds finer granularity but should never occur
+				log::debug!("RedeemState transition error: {:?}", e);
+				Error::<T>::from(RedeemError::Increase)
+			})?;
+		Pallet::<T>::apply_redeem_state_transition(who, investment_id, post_state)?;
+
+		Ok(())
+	}
+
+	#[transactional]
+	fn decrease_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+		payout_currency: T::CurrencyId,
+	) -> Result<T::Balance, DispatchError> {
+		// TODO(future): This error needs to be communicated to sending domain as it
+		// cannot be resolved by triggering a bot
+		ensure!(
+			RedemptionPayoutCurrency::<T>::get(who, investment_id)
+				.map(|currency| currency == payout_currency)
+				.unwrap_or_else(|| {
+					log::debug!("Redemption payout currency missing when calling decrease. Should never occur if redemption has been increased beforehand");
+					false
+				}),
+			Error::<T>::InvalidRedemptionPayoutCurrency
+		);
+		// TODO(future): Add implicit collection or error handling (i.e. message to
+		// source domain)
+		ensure!(
+			!T::Investment::redemption_requires_collect(who, investment_id),
+			Error::<T>::RedeemError(RedeemError::CollectRequired)
+		);
+
+		let pre_state = RedemptionState::<T>::get(who, investment_id);
+		let post_state = pre_state
+			.transition(RedeemTransition::DecreaseRedeemOrder(amount))
+			.map_err(|e| {
+				log::debug!("RedeemState transition error: {:?}", e);
+				Error::<T>::from(RedeemError::Decrease)
+			})?;
+		Pallet::<T>::apply_redeem_state_transition(who, investment_id, post_state)?;
+
+		Ok(amount)
+	}
+
+	#[transactional]
+	fn collect_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		// Note: We assume the configured Investment trait to notify about the collected
+		// amounts via the `CollectedInvestmentHook` which handles incrementing the
+		// `CollectedInvestment` amount.
+		T::Investment::collect_investment(who.clone(), investment_id)?;
+
+		Self::transfer_collected_investment(
+			who,
+			investment_id,
+			foreign_payout_currency,
+			pool_currency,
+		)
+	}
+
+	#[transactional]
+	fn collect_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		ensure!(
+			RedemptionPayoutCurrency::<T>::get(who, investment_id)
+				.map(|currency| currency == foreign_payout_currency)
+				.unwrap_or_else(|| {
+					log::debug!("Corruption: Redemption payout currency missing when calling decrease. Should never occur if redemption has been increased beforehand");
+					false
+				}),
+			Error::<T>::InvalidRedemptionPayoutCurrency
+		);
+		ensure!(T::PoolInspect::currency_for(investment_id.of_pool())
```

Comment:
I feel like these checks are too much responsibility for `foreign-investment` (and make testing harder). I think `foreign-investment` should pay focus on making the conversion and the investment, independently of what that currency means for a pool. Probably the caller of these trait's functions should ensure that for the use case of the calling context.

Reply:
In the same way, the caller should also ensure that `pool_currency` is actually a pool currency for its context.

Reply:
BTW, a question, `investment_id.of_pool()` should be always `pool_currency`?

Reply:
But should be the same, right? I mean, for an investment done in a tranche currency associated with pool A, should I make a swap order to obtain that pool currency?

Reply:
> BTW, a question, T::PoolInspect::currency_for(investment_id.of_pool()) should be always pool_currency?

Yes, if it does not return `None`. 

>  I mean, for an investment done in a tranche currency associated with pool A, should I make a swap order to obtain that pool currency?

Investments should not be possible with tranche currencies as these are the desired result of investments. As far as `pallet-investment` is concerned, investments can only be made in the pool currency. The tranche tokens are used for redemptions which will ultimately swap the tranche currency to pool currency.

Reply:
Yes! I meant for an investment using an `investment_id` represented as `(PoolA, TrancheX)`, should the `pool_currency` using in `increase_*/decrease_*` calls be a `CurrecyOfPoolA`. If this is the case, could then `pool_currency` be removed from the API interface?

Reply:
I think either way is fine. Right now, the approach is to determine the pool currency in `LP/inbound.rs` and forward it to `ForeignInvestments` because the same is done for determining the payout currencies. I think that's more straight forward than deriving the one currency in `LP/inbound.rs` and the other in the `ForeignInvestment` impl. So I would keep as is.